### DISCUSSION TITLE
Getting cheapest tariff and corresponding hour

### DISCUSSION THREAD

**Discussion by rinseringma** [fsaris/home-assistant-zonneplan-one#41] [Q&A] [answered]
Hi!

Is there a way to get the cheapest tariff and get the corresponding hour with it? Just like shown in the Zonneplan app?

Would like to expose those values to Google Assistant so I could make a routine where I ask when the cheapest hour is and what the tariff is.

**Comment by tieskuh** [comment]
The forecast prices are in an attribute of the current_electricity_tariff sensor.
Try this in the template editor:

```
{% set cheapest = state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
                  | selectattr('datetime', '>', utcnow().isoformat())
                  | sort(attribute='price')  
                  | first %}

{{ cheapest.price /10000000 }}
{{ cheapest.datetime }}
```

<img width="679" alt="image" src="https://user-images.githubusercontent.com/115901851/211219241-27f8a565-6a39-459f-9b25-c12ad0859f58.png">



You can for example create a template sensor for the cheapest upcoming hour:

```
template:
  - sensor:
    - name: zonneplan_cheapest_hour
      state: >
        {% set cheapest = state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
                    | selectattr('datetime', '>', utcnow().isoformat())
                    | sort(attribute='price')  
                    | first %}
        {{ cheapest.datetime }}
```

Hope this helps.

**Comment by rinseringma** [comment]
Thanks! But somehow the date and price aren't the same as the Zonneplan app. But that has nothing to do with your script.

Two questions;
1. Is there a way to limit the quantity of forecast hours to the upcoming, let's say, 10 hours instead of 24 hours?
2. Is it possible to format the datetime stamp to only show the hour? And maybe calculate how many hours that is from the current time?

**Comment by tieskuh** [comment, marked as answer]
The price shown in the app is the lowest price of the current day (not taken in account the prices for the next day). I think that is why it's different.

You can do anything with the data you want.

For example, this will give you the next 10 hours:

```
{{ (state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
                  | selectattr('datetime', '>', utcnow().isoformat())
                  | sort(attribute='datetime'))
                  [:10] }}
```

Or the cheapest 10 hours:

```
{{ (state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
                  | selectattr('datetime', '>', utcnow().isoformat())
                  | sort(attribute='price'))
                  [:10] }}
```

And to only get the hour from the datetume you can use time functions: https://www.home-assistant.io/docs/configuration/templating/#time
For the example in my previous comment:

```
{% set cheapest = state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
                  | selectattr('datetime', '>', utcnow().isoformat())
                  | sort(attribute='price')  
                  | first %}

{{ cheapest.price /10000000 }}
{{ as_datetime(cheapest.datetime).hour }}
```

But I think these questions are not really in the right place. This is not an issue with this integration. I think you will find most answers to these questions in the community or feel free to send me a DM.

**Reply by rinseringma** [reply]
Started to get the hang of it! I had to dig deep in my Python brain though..

Ended up with this two templates/sensors:

```
{% set first_ten_hours = (state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
| selectattr('datetime', '>', utcnow().isoformat())
| sort(attribute='datetime'))
[:10] %}

{% set cheapest = first_ten_hours 
| selectattr('datetime', '>', utcnow().isoformat())
| sort(attribute='price')  
| first %}

{{ (cheapest.price /10000000)|round(2) }}
{{ as_datetime(cheapest.datetime).strftime('%d-%m - %H:%M') }}
```

<img width="508" alt="Screenshot 2023-01-10 at 15 58 13" src="https://user-images.githubusercontent.com/2634987/211585274-bdbf60c0-2f91-4192-b4c5-3fae496ebcf7.png">

Now I need to figure out if I can expose this sensors to Google Assistant..

**Reply by tieskuh** [reply]
Great! Good luck with that! I don't use Google Assistant so I can't help you with that...

Oh and I think you should convert the time to local time at the end. The times seem to be in UTC so you are off 1 hour now.
You can also shorten your code a bit more if you'd like. 

```
{% set cheapest_hour_next_ten_hours = state_attr('sensor.zonneplan_current_electricity_tariff', 'forcast') 
  | selectattr('datetime', '>', utcnow().isoformat())
  | selectattr('datetime', '<', (utcnow() + timedelta(hours = 10)).isoformat())  
  | sort(attribute='price')  
  | first %}

{{ (cheapest_hour_next_ten_hours.price /10000000)|round(2) }}
{{ as_local(as_datetime(cheapest_hour_next_ten_hours.datetime)).strftime('%d-%m - %H:%M') }}
```

**Comment by tieskuh** [comment]
@fsaris Maybe you can enable discussions on your repository for these kind of questions

**Comment by fsaris** [comment]
Done https://github.com/fsaris/home-assistant-zonneplan-one/discussions

**Comment by TFhdKi95ae8L** [comment]
```
{% set cheapest_hour_next_twelve_hours = state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') 
  | selectattr('datetime', '>', utcnow().isoformat())
  | selectattr('datetime', '<', (utcnow() + timedelta(hours = 12)).isoformat())  
  | sort(attribute='price')  
  | first %}

{{ as_local(as_datetime(cheapest_hour_next_twelve_hours.datetime)).strftime('%d-%m - %H:%M') }}
```



For next 12 hours, and modified forcast to the now corrected for**e**cast

But how to fit this into a binary sensor? (one that can be used as a state trigger in automations, for switching on appliances)

**Reply by joopdo** [reply]
You can consider something like this:
````
sensor:
  - platform: template
    sensors:
      cheapest_hour_next_twelve_hours:
        friendly_name: "Cheapest Hour Next 12 Hours"
        value_template: >
          {% set cheapest_hour_next_twelve_hours =
            state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
            selectattr('datetime', '>', utcnow().isoformat()) |
            selectattr('datetime', '<', (utcnow() + timedelta(hours=12)).isoformat()) |
            sort(attribute='price') %}
          {% if cheapest_hour_next_twelve_hours | length > 0 %}
            {{ strptime(cheapest_hour_next_twelve_hours[0].datetime, '%Y-%m-%dT%H:%M:%S.%f%z').hour }}
          {% else %}
            N/A
          {% endif %}
        attribute_templates:
          price: >
            {% set cheapest_hour_next_twelve_hours =
              state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
              selectattr('datetime', '>', utcnow().isoformat()) |
              selectattr('datetime', '<', (utcnow() + timedelta(hours=12)).isoformat()) |
              sort(attribute='price') %}
            {% if cheapest_hour_next_twelve_hours | length > 0 %}
              {{ cheapest_hour_next_twelve_hours[0].price / 1000000 }}
            {% else %}
              N/A
            {% endif %}
````


However, I currently trigger a daily message with the following template (without a sensor):

````
{% set timezone_offset = 2 %} {# Verander 2 naar je gewenste offset in uren #}
{% set cheapest_hour_next_fifteen_hours =
state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
selectattr('datetime', '>', utcnow().isoformat()) |
selectattr('datetime', '<', (utcnow() + timedelta(hours = 15)).isoformat())
| sort(attribute='price') %}
{% if cheapest_hour_next_fifteen_hours | length > 0 %}
{% set cheapest_hour = cheapest_hour_next_fifteen_hours | first %}
{% set cheapest_hour_local_time = as_timestamp(strptime(cheapest_hour.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
De goedkoopste tijd is {{ 'vandaag' if as_timestamp(utcnow())|timestamp_custom('%Y-%m-%d') == cheapest_hour_local_time|timestamp_custom('%Y-%m-%d') else 'morgen' }} om {{ (cheapest_hour_local_time)|timestamp_custom('%H') }} uur en kost €{{"{:.2f}".format(cheapest_hour.price|float/10000000) }}/kWh.
  {% endif %}
  
  {% set cheapest_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='price') | first %}
  
  {% set expensive_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='price') | last %}
  
  {% for forecast in
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='datetime') %}
  
  {% set forecast_local_time = as_timestamp(strptime(forecast.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
  
  • {{ (forecast_local_time)|timestamp_custom('%H:%M') }}  €{{ (forecast.price / 10000000) | round(2) }}{% if forecast == cheapest_forecast %}⭐{% endif %}{% if forecast == expensive_forecast %}🔴{% endif %}
  {% endfor %}
  ````
  
  This creates the following report:
  
  ````
  De goedkoopste tijd is vandaag om 13 uur en kost €0.15/kWh.

  • 09:00  €0.24
  • 10:00  €0.2
  • 11:00  €0.18
  • 12:00  €0.15
  • 13:00  €0.15⭐
  • 14:00  €0.15
  • 15:00  €0.16
  • 16:00  €0.17
  • 17:00  €0.21
  • 18:00  €0.28  
  • 19:00  €0.31
  • 20:00  €0.32🔴
  • 21:00  €0.3
  • 22:00  €0.3
  • 23:00  €0.29
  ````

**Reply by fsaris** [reply]
@joopdo I updated your template to use the correct attribute values for price and added the price without vat

```
{% set timezone_offset = 2 %} {# Verander 2 naar je gewenste offset in uren #}
{% set cheapest_hour_next_fifteen_hours =
state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
selectattr('datetime', '>', utcnow().isoformat()) |
selectattr('datetime', '<', (utcnow() + timedelta(hours = 15)).isoformat())
| sort(attribute='price') %}
{% if cheapest_hour_next_fifteen_hours | length > 0 %}
{% set cheapest_hour = cheapest_hour_next_fifteen_hours | first %}
{% set cheapest_hour_local_time = as_timestamp(strptime(cheapest_hour.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
De goedkoopste tijd is {{ 'vandaag' if as_timestamp(utcnow())|timestamp_custom('%Y-%m-%d') == cheapest_hour_local_time|timestamp_custom('%Y-%m-%d') else 'morgen' }} om {{ (cheapest_hour_local_time)|timestamp_custom('%H') }} uur en kost €{{"{:.2f}".format(cheapest_hour.electricity_price_excl_tax|float/10000000) }}/kWh.
  {% endif %}
  
  {%- set cheapest_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='electricity_price_excl_tax') | first %}
  
  {%- set expensive_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='electricity_price_excl_tax') | last %}
  
  {%- for forecast in
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='datetime') %}
  
  {%- set forecast_local_time = as_timestamp(strptime(forecast.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
  • {{ (forecast_local_time)|timestamp_custom('%H:%M') }}    €{{ (forecast.electricity_price / 10000000) | round(2) }} (€{{ (forecast.electricity_price_excl_tax / 10000000) | round(2) }} excl.) {% if forecast == cheapest_forecast %}⭐{% endif %}{% if forecast == expensive_forecast %}🔴{% endif %}
  {%- endfor %}
```

![Screenshot_20250604-082604 Home Assistant](https://github.com/user-attachments/assets/ed15ce4c-e00c-42cf-be89-545fd725a884)

**Reply by tshw3** [reply]
> @joopdo I updated your template to use the correct attribute values for price and added the price without vat


Hi all,

I'm really happy with this code, it works perfectly in Developer Tools → Template.
However, I'm a bit stuck and sorry for the beginner question: how do I add this to my dashboard?

Do I need to include it in configuration.yaml?
When I try that, I get errors because it’s not valid YAML code. I even tried converting it using ChatGPT, but that didn’t work either.
What am I doing wrong?

This is probably basic knowledge, but here’s my process:

1) I test something in Developer Tools → Template Editor
2) I want to use it on the dashboard — what’s the next step?

Thanks so much for your help.

Also, I really appreciate the work you’ve put into this HACS plugin. It works great!

**Reply by fsaris** [reply]
ah there is a small mistake  in the example, will update the readme. You can use this template in a markdown card like:

```
type: markdown
content: >-

  {% set timezone_offset = 2 %} {# Verander 2 naar je gewenste offset in uren #}

  {% set cheapest_hour_next_fifteen_hours =

  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |

  selectattr('datetime', '>', utcnow().isoformat()) |

  selectattr('datetime', '<', (utcnow() + timedelta(hours = 15)).isoformat())

  | sort(attribute='electricity_price') %}

  {% if cheapest_hour_next_fifteen_hours | length > 0 %}

  {% set cheapest_hour = cheapest_hour_next_fifteen_hours | first %}

  {% set cheapest_hour_local_time =
  as_timestamp(strptime(cheapest_hour.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) +
  timezone_offset * 3600 %}

  De goedkoopste tijd is {{ 'vandaag' if
  as_timestamp(utcnow())|timestamp_custom('%Y-%m-%d') ==
  cheapest_hour_local_time|timestamp_custom('%Y-%m-%d') else 'morgen' }} om {{
  (cheapest_hour_local_time)|timestamp_custom('%H') }} uur en kost
  €{{"{:.2f}".format(cheapest_hour.electricity_price_excl_tax|float/10000000)
  }}/kWh.
    {% endif %}
    
    {%- set cheapest_forecast =
    state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
    selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
    '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
    sort(attribute='electricity_price_excl_tax') | first %}
    
    {%- set expensive_forecast =
    state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
    selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
    '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
    sort(attribute='electricity_price_excl_tax') | last %}
    
    {%- for forecast in
    state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
    selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
    '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
    sort(attribute='datetime') %}
    
    {%- set forecast_local_time = as_timestamp(strptime(forecast.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
    • {{ (forecast_local_time)|timestamp_custom('%H:%M') }}    €{{ (forecast.electricity_price / 10000000) | round(2) }} (€{{ (forecast.electricity_price_excl_tax / 10000000) | round(2) }} excl.) {% if forecast == cheapest_forecast %}⭐{% endif %}{% if forecast == expensive_forecast %}🔴{% endif %}
    {%- endfor %}
```

**Reply by tshw3** [reply]
Hi Fsaris,

Really appreciate your fast reply, and this one works great! 

Have a nice evening.

**Reply by stiffupperlip13** [reply]
I'm trying to implement this code in developer tools but I keep getting this error: TypeError: '>' not supported between instances of 'datetime.datetime' and 'str'. Did something change in HA or is it me? I'm a beginner and still learning. Would appreciate your help. 

Im using this code:
{% set timezone_offset = 2 %} {# Verander 2 naar je gewenste offset in uren #}
{% set cheapest_hour_next_fifteen_hours =
state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
selectattr('datetime', '>', utcnow().isoformat()) |
selectattr('datetime', '<', (utcnow() + timedelta(hours = 15)).isoformat())
| sort(attribute='price') %}
{% if cheapest_hour_next_fifteen_hours | length > 0 %}
{% set cheapest_hour = cheapest_hour_next_fifteen_hours | first %}
{% set cheapest_hour_local_time = as_timestamp(strptime(cheapest_hour.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
De goedkoopste tijd is {{ 'vandaag' if as_timestamp(utcnow())|timestamp_custom('%Y-%m-%d') == cheapest_hour_local_time|timestamp_custom('%Y-%m-%d') else 'morgen' }} om {{ (cheapest_hour_local_time)|timestamp_custom('%H') }} uur en kost €{{"{:.2f}".format(cheapest_hour.electricity_price_excl_tax|float/10000000) }}/kWh.
  {% endif %}
  
  {%- set cheapest_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='electricity_price_excl_tax') | first %}
  
  {%- set expensive_forecast =
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='electricity_price_excl_tax') | last %}
  
  {%- for forecast in
  state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') |
  selectattr('datetime', '>', utcnow().isoformat()) | selectattr('datetime',
  '<', (utcnow() + timedelta(hours = 15)).isoformat()) | list |
  sort(attribute='datetime') %}
  
  {%- set forecast_local_time = as_timestamp(strptime(forecast.datetime, '%Y-%m-%dT%H:%M:%S.%fZ')) + timezone_offset * 3600 %}
  • {{ (forecast_local_time)|timestamp_custom('%H:%M') }}    €{{ (forecast.electricity_price / 10000000) | round(2) }} (€{{ (forecast.electricity_price_excl_tax / 10000000) | round(2) }} excl.) {% if forecast == cheapest_forecast %}⭐{% endif %}{% if forecast == expensive_forecast %}🔴{% endif %}
  {%- endfor %}

**Reply by fsaris** [reply]
@stiffupperlip13 did you try the version from the readme?

**Reply by stiffupperlip13** [reply]
> @stiffupperlip13 did you try the version from the readme?
When I change the sensor to: sensor.zonneplan_current_quarter_hourly_electricity_tariff all over the code it does give me output. However it now gives quarterly prices so this may mean it needs some rewriting. 
As I said i'm new to this but eager to learn. Can you confirm we indeed need to change the sensor? Or is it something i'm doing wrong?

Also the cheapest time is not right. it says 0,021/kwh whiile in fact it's 0,014. Maybe it has something to do with the quarter/hour conversion in the code.

Thanks in advance for your reply! You can also dm me if you like.

**Reply by stiffupperlip13** [reply]
Edit: the cheapest time is right. It tells me the price excl. I was confused about that but I get it now. I still like to convert the code into either hourly prices or quarterly prices.

**Reply by fsaris** [reply]
> I still like to convert the code into either hourly prices or quarterly prices.

What are you looking for?

**Reply by stiffupperlip13** [reply]
> What are you looking for?

Every day at a certain time I want to send a notification to my phone of the cheapest hour on that day (or maybe also the 2nd cheapest). My house runs on electricity only with a dynamic contract. I'd also like to trigger some of my devices at the cheapest hour, like the washing machine.

Could you tell me how I achieve those things? Currently I made a sensor template with (part of) your code. I had to cut a part out because it told me the code string was too long. I then made an automation that sends the state of the sensor to my phone every morning. Next thing I want to achieve is to run the washing machine at the cheapest hour. I can set the machine every morning. I integrated it in HA and i'm able to trigger 'start program'.

Long story short. I'm a 'noob' eager to learn code. I'm experienced with Excel and SQL but this is all new to me.

**Reply by fsaris** [reply]
@stiffupperlip13 I just created a blueprint. Maybe that can help you with your task. See the [blueprints folder](https://my.home-assistant.io/redirect/blueprint_import/?blueprint_url=https%3A%2F%2Fgithub.com%2Ffsaris%2Fhome-assistant-zonneplan-one%2Fblob%2Fadd-blueprint-for-cheapest-price-window%2Fblueprints%2Fzonneplan_cheapest_window.yaml) or preview/import it in you own HA instance https://my.home-assistant.io/redirect/blueprint_import/?blueprint_url=https%3A%2F%2Fgithub.com%2Ffsaris%2Fhome-assistant-zonneplan-one%2Fblob%2Fadd-blueprint-for-cheapest-price-window%2Fblueprints%2Fzonneplan_cheapest_window.yaml

**Reply by stiffupperlip13** [reply]
@fsaris 

Thank you for this! I actually had my own template running (with your code) until this morning. I'm getting TypeError: '>' not supported between instances of 'datetime.datetime' and 'str'. The blueprint also won't work anymore. Maybe the YAML settings have been changed in the latest update.

**Reply by fsaris** [reply]
Are you using the new forecast sensor?

**Reply by stiffupperlip13** [reply]
I got my code to work again. Turns out I had to remove "isoformat" out of the yaml code. As I said, yesterday it worked fine so maybe some yaml settings changed during the last HA update.

I also got the blueprint imported. Thanks a lot for that. Now I can actually turn on devices when the power is cheap.

**Reply by fsaris** [reply]
> Turns out I had to remove "isoformat" out of the yaml code. 

This was changed in the latest release of the Zonneplan integration. De the breaking change notice in the release notes

**Reply by stiffupperlip13** [reply]
Did it affect the blueprint as well? It gives the following error when executed: 
Variabelen best_window definiëren
Uitgevoerd op: 9 augustus 2026 om 19:45:00
Fout: UndefinedError: 'dict object' has no attribute 'end_date'
Sjabloonfouten:

    'dict object' has no attribute 'end_date'

**Reply by stiffupperlip13** [reply]
Got it to work by using the quarterly tarrif. This is great work, thanks again!

**Comment by Robbedoes2000** [comment]
Well, I don't know if I should ask this question here, but I want to switch a charger on at for example the lowest price hour. However, I'm not familiar with python and don't understand yaml yet. Anyone already did something like this?

**Comment by ronaldtveen** [comment]
@fsaris this example in the docs should also be updated:

```
{% set cheapest_hour_next_twelve_hours = state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') 
  | selectattr('datetime', '>', utcnow().isoformat())
  | selectattr('datetime', '<', (utcnow() + timedelta(hours = 11)).isoformat())  
  | sort(attribute='price')  
  | first %}

{{ as_local(as_datetime(cheapest_hour_next_twelve_hours.datetime)) }}
```

the attribute `price` should actually be `electricity_price` for it to work correctly:

```
{% set cheapest_hour_next_twelve_hours = state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') 
  | selectattr('datetime', '>', utcnow().isoformat())
  | selectattr('datetime', '<', (utcnow() + timedelta(hours = 11)).isoformat())  
  | sort(attribute='electricity_price')  
  | first %}

{{ as_local(as_datetime(cheapest_hour_next_twelve_hours.datetime)) }}
```

**Reply by fsaris** [reply]
You are correct, fixed. See also #194

**Reply by Thephaseshifter** [reply]
Hi fsaris, i was able to install your Zonneplan integration and the sensor helper for the price-graph
Also the "laagste prijs tijdstip" sensor readout and graph both shows fine in my dashbord
But i am strugeling to automate the EVCS to start on a trigger that can be based on that "laagste prijs tijdstip"
Can you help me to provide a set-trigger for my EVCS ?
Dashbord EVCS showing date/time format "sensor.laagste prijs tijdstip"
<img width="1663" height="881" alt="EVCS" src="https://github.com/user-attachments/assets/c5b91fac-c58c-4bd0-9d8d-b519f19ffefa" />

**Reply by fsaris** [reply]
What would be the trigger and requirements to start charging?

1. Connected to charger
2. Set to manual charging
3. Cheapest hour (in what timeframe?)
4. ....

Should it only charge for 1hour?


Did you see https://github.com/fsaris/home-assistant-zonneplan-one/discussions/215 ?

**Reply by Thephaseshifter** [reply]
Hi Frans, thanks for your swift reply and your effort on writing the zonneplan integration!

The automation for handling the EVCS itself is clear to me, tested it with a simple trigger and works.
I only do not understand how to use the "laagste prijs tijdstip" within a 12 hour timeframe. It seems a sensor value with date&time value but i can't use it or do not know how to use as a binary "transition from 0 to 1" trigger to set a automation.

**Reply by Thephaseshifter** [reply]
Hi Frans, at this moment i noticed that the helper-template - template option - device class, was not set properly (overlooked it), so just selected "Time" 
Now waiting for the automation to set to ON on that trigger...

**Reply by Thephaseshifter** [reply]
Today the first start would be at 14:00 hour but than at 13:00 hours, the Zonneplan pricing showed a lower price at 00:00 time, so no trigger to start my automation was engaged.
Copied the original Template (11 hour ahead range) Changed to 8 hour ahead and added that copie as a second trigger for my automation, so now wait and see if start will be innitiated at 14:00 

I changed this in the Template:
{% set cheapest_hour_next_**nine**_hours = state_attr('sensor.zonneplan_current_electricity_tariff', 'forecast') 
  | selectattr('datetime', '>', utcnow().isoformat())
  | selectattr('datetime', '<', (utcnow() + timedelta(hours = **8**)).isoformat())  
  | sort(attribute='electricity_price')  
  | first %}

{{ as_local(as_datetime(cheapest_hour_next_twelve_hours.datetime)) }}
.


alias: "EVCS Zonneplan-start "
description: >-
  Het Victron EV-laadstation zal starten zodra datum&tijdstempel van de template
  "laagste prijs zonneplan" is bereikt
triggers:
  - trigger: time
    at: sensor.laagste_prijs_volgende_12_uur
  - trigger: time
    at: sensor.template_zonneplan_lp
conditions:
  - type: is_on
    condition: device
    device_id: cd4fcb39229681cd7d29a31af64bbecd
    entity_id: 59f52f0cfcd10b2b768748d143b948df
    domain: binary_sensor
  - condition: not
    conditions:
      - condition: device
        device_id: cd4fcb39229681cd7d29a31af64bbecd
        domain: select
        entity_id: ff25945dff4adb47608eac29c97f5af2
        type: selected_option
        option: Scheduled Charge
actions:
  - device_id: cd4fcb39229681cd7d29a31af64bbecd
    domain: select
    entity_id: ff25945dff4adb47608eac29c97f5af2
    type: select_option
    option: Manual
  - type: turn_on
    device_id: cd4fcb39229681cd7d29a31af64bbecd
    entity_id: e9f8e76f1737433dd2b5b6471c114835
    domain: switch
  - action: notify.mobile_app_iphone_rene
    metadata: {}
    data:
      message: >-
        De Victron EVCS is gestart op het laagste prijs moment van het Zonneplan
        kontrakt !!
  - delay:
      hours: 3
      minutes: 0
      seconds: 0
      milliseconds: 0
  - type: turn_off
    device_id: cd4fcb39229681cd7d29a31af64bbecd
    entity_id: e9f8e76f1737433dd2b5b6471c114835
    domain: switch
  - action: notify.mobile_app_iphone_rene
    metadata: {}
    data:
      message: De start op Zonneplan voorwaarde voor de Victron EVCS is gestopt.
mode: restart

**Reply by fsaris** [reply]
With `selectattr('datetime', '<', (utcnow() + timedelta(hours = 8)).isoformat())` you have a constant sliding value sensor. You need to one of the value fixed.

Some ideas:
1. Create a automation that sets new next charge time sensor based on your already existing helper
2. Change the delta mentioned above to a fix date. For instance charger connect time+ 9hours.

**Reply by Thephaseshifter** [reply]
Good morning Frans,
Writing code such as Python is not in my skills, i am more used to plc logic like stepladder (as pensionado)
Especially date/time format gets me nervous ;-)
Anyway, the automation works fine on the 9 hours ahead, so i'll erase the 12 h prediction timer for the time being.
Waiting for Victron community to come with a solution.

Thanks for your assistance!

**Comment by sanderdewit** [comment]
sharing my markdown, as I didn't want to update the column every time it's summer/winter time.
Also made sure the 0,1 because 0,10 etc. Changed dots into comma's to compile with European logic.

```
{% set fc = state_attr('sensor.zonneplan_current_electricity_tariff','forecast') | default([]) 
  | selectattr('datetime','>', utcnow().isoformat())
  | selectattr('datetime','<', (utcnow() + timedelta(hours=15)).isoformat())
  | list %}

{% if fc | length > 0 %}
{% set cheapest_hour = (fc | sort(attribute='electricity_price') | list | first) %}
{% set cheapest_hour_local_time = as_local(as_datetime(cheapest_hour.datetime)) %}
De goedkoopste tijd is {{ 'vandaag' if cheapest_hour_local_time.date() == now().date() else 'morgen' }} om {{ cheapest_hour_local_time.strftime('%H:%M') }} uur en kost
€{{ "{:.2f}".format(cheapest_hour.electricity_price | float / 10000000).replace('.', ',') }}/kWh.
{% endif %}

{%- set sorted_by_excl = (fc | sort(attribute='electricity_price_excl_tax') | list) %}
{%- set cheapest_forecast = sorted_by_excl | first %}
{%- set expensive_forecast = sorted_by_excl | last %}
{%- for forecast in (fc | sort(attribute='datetime') | list) %}
{%- set forecast_local_time = as_local(as_datetime(forecast.datetime)) %}
    • {{ forecast_local_time.strftime('%H:%M') }}    €{{ "{:.2f}".format(forecast.electricity_price | float / 10000000).replace('.', ',') }} (€{{ "{:.2f}".format(forecast.electricity_price_excl_tax | float / 10000000).replace('.', ',') }} excl.){%- if forecast == cheapest_forecast -%}⭐{%- endif -%}{%- if forecast == expensive_forecast -%}🔴{% endif %}
{%- endfor %}
```
